extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java

```diff
@@ -0,0 +1,93 @@
+            DockerStatusBuildItem dockerStatusBuildItem) {
+        if (ConfigUtils.isPropertyNonEmpty(OIDC_CLIENT_REG_AUTH_SERVER_URL_CONFIG_KEY)) {
+            LOG.debugf("Not starting Dev Services for Keycloak as '%s' has been provided",
+                    OIDC_CLIENT_REG_AUTH_SERVER_URL_CONFIG_KEY);
+            return;
+        }
+        // this is for backwards compatibility - when users configured named client with OIDC auth server URL
+        // that is not our Keycloak Dev Service, we don't want to start our Dev Service to keep previous behavior
+        if (ConfigUtils.isPropertyNonEmpty(OIDC_AUTH_SERVER_URL_CONFIG_KEY)) {
+            LOG.debugf("Not starting Dev Services for Keycloak as '%s' has been provided",
+                    OIDC_AUTH_SERVER_URL_CONFIG_KEY);
+            return;
+        }
+        if (!dockerStatusBuildItem.isContainerRuntimeAvailable()) {
+            LOG.warnf("Please configure '%s' or get a working docker instance", OIDC_CLIENT_REG_AUTH_SERVER_URL_CONFIG_KEY);
+            return;
+        }
+
+        var devServicesConfigurator = new KeycloakDevServicesConfigurator() {
+
+            @Override
+            public Map<String, String> createProperties(ConfigPropertiesContext ctx) {
+                return Map.of(OIDC_CLIENT_REG_AUTH_SERVER_URL_CONFIG_KEY, ctx.authServerInternalUrl());
+            }
+
+            @Override
+            public void customizeDefaultRealm(RealmRepresentation realmRepresentation) {
+                if (getInitialToken() == null) {
+                    realmRepresentation.setRegistrationAllowed(true);
+                    realmRepresentation.setRegistrationFlow("registration");
+                    if (realmRepresentation.getComponents() == null) {
+                        realmRepresentation.setComponents(new MultivaluedHashMap<>());
+                    }
+                    var componentExportRepresentation = new ComponentExportRepresentation();
+                    componentExportRepresentation.setName("Full Scope Disabled");
+                    componentExportRepresentation.setProviderId("scope");
+                    componentExportRepresentation.setSubType("anonymous");
+                    realmRepresentation.getComponents().put(
+                            "org.keycloak.services.clientregistration.policy.ClientRegistrationPolicy",
+                            List.of(componentExportRepresentation));
+                }
+            }
+        };
+
+        keycloakDevSvcRequiredProducer
+                .produce(new KeycloakDevServicesRequiredBuildItem(devServicesConfigurator));
+    }
+
+    @BuildStep(onlyIf = IsDevelopment.class)
+    KeycloakAdminPageBuildItem addCardWithLinkToKeycloakAdmin() {
+        return new KeycloakAdminPageBuildItem(new CardPageBuildItem());
+    }
+
+    private static String getInitialToken() {
+        return ConfigProvider.getConfig().getOptionalValue("quarkus.oidc-client-registration.initial-token", String.class)
```

Comment:
Does it really work though, it is a runtime property

Reply:
I don't have initial token, so if you don't mind, I'll answer with another question. Not sure it's fair, but:

Quarkus OIDC auth server URL property https://quarkus.io/guides/all-config#quarkus-oidc_quarkus-oidc-auth-server-url is checked before this code is executed https://github.com/quarkusio/quarkus/blob/main/extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakDevServicesProcessor.java#L369, while this code you are pointing to is executed after KC containers has started and you are creating default realm. What is difference? Either all runtime property can be queries or none.

Reply:
@sberyozkin I have played with a dev mode for a while now and yes, it works (though I think the answer above explain the reason why it is expect to work as well). so what happens is that when initial token is set (I did set nonsense), KC is not configured with anonymous policy and default client registration fails. You would need to configure KC with realm.json as you did inside of the OIDC client integration test.

Reply:
I also tried to apply step by step:

```
quarkus.oidc-client.enabled=false
quarkus.oidc-client-registration.enabled=false
quarkus.oidc.enabled=false
```
and enable / disable individual extensions, all works as I expected. when all 3 are disabled, KC dev service is closed. when you enable at least one of them, KC is started again.